### PR TITLE
Don't set ambient caps without inheritable ones

### PR DESCRIPTION
**Pull request by kolyshkin**
Commit 98fe566c removed setting inheritable capabilities from
 - `runc exec --cap`;
 - example spec (used by `runc spec`);
 - libcontainer/integration test config;
 - libcontainer/README.md;
but neglected to also remove ambient capabilities.

An ambient capability could only be set if the same inheritable
capability is set, so as a result of the above change ambient
capabilities were not set (but due to a bug in gocapability package,
those errors are never reported).
                                                                   
Once we start using a package [with the fix][1], that bug will become
apparent (both bats-based and inct/int tests will fail).

**This PR removes setting ambient capabilities where inheritable ones
are not set. This has no effect in the current codebase (since those
ambient capabilities were not set anyway, and the errors were ignored),
but will prevent errors once we switch to the fixed capability package.**

As we do not have any tests for `runc exec --cap`, add one.

[1]: https://github.com/kolyshkin/capability/pull/3

### CODE REVIEW

File: exec.go
Comment (rata) [inline exec.go:237]: I understand that we should remove ambient if we don't set inheritable capabilities. But why we are not setting them? The commit that removed them doesn't say anything.

It will be great for the reason to remove them to be written down somewhere, at least in this commit message or PR description.

**Review by rata — APPROVED** [review]
LGTM, thanks!

However, IMHO, it would be great to document why we removed _inheritable capabilities_ in the first place. It is clear that if we remove inh, we should remove ambient too.

**Comment by kolyshkin** [comment]
> However, IMHO, it would be great to document why we removed _inheritable capabilities_ in the first place. It is clear that if we remove inh, we should remove ambient too.

All I remember it was a security fix, originally reported by cap maintainers (Andrew Morgan). Something along the lines that inheritable capabilities are usually being set from the file system attributes, and are not expected to be set this way (`runc exec --cap`).

File: exec.go
Comment (rata) [reply exec.go:237]: Marking as solved, as you answer this already in a PR comment.

File: libcontainer/README.md
Comment (lifubang) [inline libcontainer/README.md:109]: This is the example of container creation, not about exec to a container.
Yes, we can drop some cap list to simplify the example, but I think it will be better to add back the inheritable list and keep the ambient list.

File: libcontainer/specconv/example.go
Comment (lifubang) [inline libcontainer/specconv/example.go:44]: The same here.

**Comment by lifubang** [comment]
Another problem, maybe there is a situation we should consider:
A user wants to exec with a cap in inheritable cap list in ‘config.json’ but not in ambient cap list in ‘config.json’. In my brain, when doing exec, the process’s inheritable cap list is inherited from ‘config.json’, so it maybe exists in ‘runc exec’. The original security patch is to prevent extend the inheritable cap list, but not delete it.

If this is true, I think this PR is wrong.

File: libcontainer/specconv/example.go
Comment (kolyshkin) [reply libcontainer/specconv/example.go:44]: This one is not an example, but the data used by `runc spec` (and thus in all runc integration tests).

In the other comment you say it's better to turn on inheritable capabilities, but the whole point of https://github.com/opencontainers/runc/security/advisories/GHSA-f3fp-gc8g-vw66 was to NOT set inheritable capabilities by default. By extension, ambient capabilities should not be set by default either (as they won't work without inheritable ones).

**Comment by kolyshkin** [comment]
> A user wants to exec with a cap in inheritable cap list in ‘config.json’ but not in ambient cap list in ‘config.json’.

With a single `--caps` option which does not allow a user to specify the type of capabilities it's not possible. If needed, a user can use `runc exec --process process.json` and specify all the individual types of caps.

@lifubang try to look at this PR in the following way:
 - it is not possible to set ambient caps without inheritable caps set;
 - inheritable caps should not be set by default, and were removed to fix https://github.com/opencontainers/runc/security/advisories/GHSA-f3fp-gc8g-vw66;
 - without inheritable caps set, ambient caps bits are now ignored (due to a bug in gocapability);

So what this PR does is removes ambient caps which are not being set anyway (they are ignored). In other words, it changes nothing.

**Comment by lifubang** [comment]
> * inheritable caps should not be set by default, and were removed to fix [GHSA-f3fp-gc8g-vw66](https://github.com/opencontainers/runc/security/advisories/GHSA-f3fp-gc8g-vw66);

So, what you mean is that the ‘inheritable caps’ should always be empty in ‘config.json’?

**Comment by kolyshkin** [comment]
> > * inheritable caps should not be set by default, and were removed to fix [GHSA-f3fp-gc8g-vw66](https://github.com/opencontainers/runc/security/advisories/GHSA-f3fp-gc8g-vw66);
> 
> So, what you mean is that the ‘inheritable caps’ should always be empty in ‘config.json’?

Probably not always, but by default they should be empty.

Whence, ambient caps should also be empty by default, since it's not possible to set ambient without inheritable.

Hope that clears things up.

File: exec.go
Comment (lifubang) [inline exec.go:237]: I think we should add a check here:
```
if p.Capabilities.Inheritable != nil {
  p.Capabilities.Ambient = append(p.Capabilities.Ambient, c)
}
```
Or else it will cause a compatibility issue.

Furthermore, maybe the patch of `[GHSA-f3fp-gc8g-vw66](https://github.com/opencontainers/runc/security/advisories/GHSA-f3fp-gc8g-vw66)` also should be like this:
```
if p.Capabilities.Inheritable != nil {
  p.Capabilities.Inheritable = append(p.Capabilities.Inheritable, c)
  p.Capabilities.Ambient = append(p.Capabilities.Ambient, c)
}
```
But this is not very important, because It depends on our understanding of `runc exec --cap`: `add a capability to the bounding set for the process`. The important one is the compatibility issue metioned above.

Comment (kolyshkin) [reply exec.go:237]: Thanks for raising this!

Since [runc 1.1.2](https://github.com/opencontainers/runc/releases/tag/v1.1.2) we do not set inheritable capabilities in `runc exec --cap` (and it also means we do not set any ambient capabilities either), and we have yet to receive a bug report.

So perhaps this is now normal?

Comment (lifubang) [reply exec.go:237]: > the documentation also says `runc exec --caps` adds a cap to the bounding set.

How do you understand this sentence? Only add to bounding set?



> I think we should add a check here:
> 
> ```
> if p.Capabilities.Inheritable != nil {
>   p.Capabilities.Ambient = append(p.Capabilities.Ambient, c)
> }
> ```
> 
> Or else it will cause a compatibility issue.

If you can add this check to ensure we can be compatible with previous, I think it will be better.

Comment (kolyshkin) [reply exec.go:237]: Adding this test here.

Comment (kolyshkin) [reply exec.go:237]: Added the test case from #4406 and the conditional setting of ambient caps in runc exec --caps.

**Review by lifubang — APPROVED** [review]
LGTM
Waiting #4412 merged to unblock CI.

**Pull request merged**
